### PR TITLE
MLPAB-1950 - Build index from DynamoDB export

### PR DESCRIPTION
**Pull request by andrewpearce-digital**
# Purpose

The team require a method to create an Opensearch index from scratch in the event that it is lost.

Fixes MLPAB-1950

## Approach

Use an Opensearch Ingestion Service Pipeline to create an encrypted export of the DynamoDB table and process the items into index entries.

- Allow pipeline use of customer managed keys for DynamoDB and the DynamoDB exports S3 bucket
- Allow pipeline access to the DynamoDB exports S3 bucket
- ensure role, access policy and security group are in place before creating pipeline
- add a table export as a source for the ingestion pipeline
- disable app and event-received search indexing
- enable pipeline indexing

~- TODO: address first HeadBucket requests failing~ raised ticket for this https://opgtransform.atlassian.net/browse/MLPAB-2216

## Learning

- https://opensearch.org/docs/latest/data-prepper/pipelines/configuration/processors/key-value/
- https://opensearch.org/docs/latest/data-prepper/pipelines/configuration/sources/dynamo-db/
- https://opensearch.org/docs/latest/data-prepper/pipelines/configuration/sinks/opensearch/

### CODE REVIEW

File: terraform/environment/opensearch_ingestion_pipeline.tf [hunk 1]
```diff
@@ -30,7 +39,10 @@ data "aws_iam_policy_document" "opensearch_pipeline" {
       "aoss:BatchGetCollection",
       "aoss:APIAccessAll"
     ]
-    resources = [data.aws_opensearchserverless_collection.lpas_collection.arn]
+    resources = [
+      "*",
```
Comment (andrewpearce-digital) [inline terraform/environment/opensearch_ingestion_pipeline.tf:43]: remove this

File: terraform/environment/opensearch_ingestion_pipeline.tf [hunk 2]
```diff
@@ -56,21 +68,34 @@ data "aws_iam_policy_document" "opensearch_pipeline" {
...
 
   statement {
-    sid    = "DynamoDBEncryptionAccess"
+    sid    = "DescribeExports"
+    effect = "Allow"
+    actions = [
+      "dynamodb:DescribeExport",
+    ]
+    resources = [
+      "${aws_dynamodb_table.lpas_table.arn}/export/*",
```
Comment (andrewpearce-digital) [inline terraform/environment/opensearch_ingestion_pipeline.tf:85]: this is no longer the path to exports

File: terraform/environment/opensearch_ingestion_pipeline.tf [hunk 3]
```diff
@@ -84,10 +109,10 @@ data "aws_iam_policy_document" "opensearch_pipeline" {
     ]
     resources = [
       data.aws_kms_alias.opensearch_encryption_key.target_key_arn,
```
Comment (andrewpearce-digital) [inline terraform/environment/opensearch_ingestion_pipeline.tf:111]: remove this

File: docs/architecture/decisions/0005-namespacing-resources-in-aws.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+(this leads to the consequence that resources will have either `environment-name` or `account-name` and not likely to have both)
+
+## Consequences
+
+- The app service will stop putting items into the index directly
+- DynamoDB streeam will be enabled for all environments
+- An ingestion pipeline will be responsible for adding items to the index from
+  the DynamoDB event stream
+- An ingestion pipeline will be responsible for recovering the index from a
+  table export if it is lost
```
Comment (acsauk) [inline docs/architecture/decisions/0005-namespacing-resources-in-aws.md:76]: Is this copy pasta from the previous ADR?

Comment (andrewpearce-digital) [reply docs/architecture/decisions/0005-namespacing-resources-in-aws.md:76]: oh, yeah it is. I'll get that sorted

**Review by acsauk — APPROVED** [review]
:shipit:

**Pull request merged**